gordo/cli/cli.py

```diff
@@ -38,7 +38,6 @@
     NoSuitableDataProviderError: 70,
     InsufficientDataError: 80,
     InsufficientDataAfterRowFilteringError: 81,
-    MlflowLoggingError: 90,
```

Comment:
Removing this import makes sense; however, having the mappings all in one place also makes sense. Talked with Erik about it, and his suggestion is to move the reporter exeception(s) to their own file/module adjacent to the reporter modules (i.e. `gordo/reporters/exceptions`, which then can be imported by both the reporter and cli modules.

Reply:
cc: @epa095

Ok. 

This implies that anyone also implementing a reporter needs to somehow get their exception classes into this file, and in this respect, I see nothing wrong with updating the mapping within the respective reporters; that would be the natural way to go about it in practice.

Atop of that, we'll then essentially be importing and mapping error types to codes for a feature which may not even be installed in the first place?

Reply:
> This implies that anyone also implementing a reporter needs to somehow get their exception classes into this file

Yes, and they need to modify `gordo/cli/cli.py` to map it to a exit code. I am also fine with having the mlflow exception in a file by itself, if it is more natural to not have the exceptions in a common file. 

I don't think it is good for the ownership of `EXCEPTION_TO_EXITCODE` to be scattered around the codebase. It is good to be able to go to the mapping and immediately see the exit codes and what errors cause them. Having the map being updated from around the codebase makes it hard to know the complete mapping, and even opens the possibility for crashing exit codes. 

Part of the API of the reporters is what exceptions it raises, but not how the `cli` is supposed to handle that exception, that is for the cli to determine. We actually thought about another alternative earlier, to have the exception classes have a class constant `ERROR_CODE`, which contained the error code, instead of using the mapping. But it was rejected exactly because the exit code is not owned by the exception. 

One compromise I am willing to consider is the following:
If the `cli` gets a error it does not recognise in `EXCEPTION_TO_EXITCODE` it will see if it has a variable/method `recommended_exit_code`, **and** if this is between 100 and 120 it will exit with that code. Then we allow our non-existing future plugins to define their own exit codes, but we restrict them to the "unorganized region" of 100 to 120 where anything can happen.

Reply:
> Yes, and they need to modify gordo/cli/cli.py to map it to a exit code.
This isn't possible unless they fork the repo though, and we'd expect users to maintain a fork when using Gordo?

> I don't think it is good for the ownership of `EXCEPTION_TO_EXITCODE` to be scattered around the codebase.
I would contend it's not scattered, as in this context, only `reporters` are going to be updating it. Or perhaps I misunderstand.

> Part of the API of the reporters is what exceptions it raises, but not how the cli is supposed to handle that exception, that is for the cli to determine. 

In that sense, the API is broken. If the CLI is to determine how to handle exceptions raised by arbitrary reporters, that won't scale out to other implementations. 

With all that said..  :man_shrugging: I think you're more set on this than I am, so I'll yield.